google/cloud/internal/async_read_write_stream_auth.h

```diff
@@ -28,79 +29,97 @@ namespace internal {
+          if (auto state = weak.lock()) return state->OnStart(f.get());
           return make_ready_future(false);
         });
   }
 
   future<absl::optional<Response>> Read() override {
-    if (!stream_) return make_ready_future(absl::optional<Response>{});
-    return stream_->Read();
+    std::lock_guard<std::mutex> g{state_->mu};
+    return state_->stream->Read();
   }
 
   future<bool> Write(Request const& request,
                      grpc::WriteOptions options) override {
-    if (!stream_) return make_ready_future(false);
-    return stream_->Write(request, std::move(options));
+    std::lock_guard<std::mutex> g{state_->mu};
+    return state_->stream->Write(request, std::move(options));
   }
 
   future<bool> WritesDone() override {
-    if (!stream_) return make_ready_future(false);
-    return stream_->WritesDone();
+    std::lock_guard<std::mutex> g{state_->mu};
+    return state_->stream->WritesDone();
   }
 
   future<Status> Finish() override {
-    if (!stream_) {
-      return make_ready_future(
-          Status(StatusCode::kInvalidArgument,
-                 "uninitialized GrpcReadWriteStreamAuth<>"));
-    }
-    return stream_->Finish();
+    return state_->Finish();
   }
 
  private:
-  future<bool> OnStart(StatusOr<std::unique_ptr<grpc::ClientContext>> context) {
-    if (!context) {
-      stream_ =
-          absl::make_unique<AsyncStreamingReadWriteRpcError<Request, Response>>(
+  struct SharedState {
+    SharedState(StreamFactory factory, std::unique_ptr<grpc::ClientContext> initial_context) : factory(std::move(factory)), initial_context(std::move(initial_context)), stream(absl::make_unique<AsyncStreamingReadWriteRpcError<Request, Response>>(Status(StatusCode::kFailedPrecondition, "Stream is not yet started."))) {}
+
+    future<bool> OnStart(StatusOr<std::unique_ptr<grpc::ClientContext>> context) {
+      std::lock_guard<std::mutex> g{mu};
+      if (cancelled) return make_ready_future(false);
+      if (context) {
+        stream = factory(*std::move(context));
+      } else {
+        stream = absl::make_unique<AsyncStreamingReadWriteRpcError<Request, Response>>(
               std::move(context).status());
-      return make_ready_future(false);
+      }
+      return stream->Start();
     }
-    stream_ = factory_(*std::move(context));
-    return stream_->Start();
-  }
 
-  std::unique_ptr<grpc::ClientContext> context_;
-  std::shared_ptr<GrpcAuthenticationStrategy> auth_;
-  StreamFactory factory_;
-  std::shared_ptr<AsyncStreamingReadWriteRpc<Request, Response>> stream_;
+    future<Status> Finish() {
+      std::lock_guard<std::mutex> g{mu};
+      cancelled = true;  // ensure stream is not recreated after Finish
+      return stream->Finish();
+    }
+
+    void Cancel() {
+      std::lock_guard<std::mutex> g{mu};
+      if (cancelled) return;
+      cancelled = true;
+      if (initial_context) initial_context->TryCancel();
+      stream->Cancel();
+    }
+
+    const StreamFactory factory;
```

Comment:
east const or no `const`.  I am not sure const-qualifying this adds enough value 🤷

Reply:
Done. It does, because it explains why it is not mutex guarded: its value never changes.